### PR TITLE
Cleanup make test-integration

### PR DESCRIPTION
**Pull request by marun**
``make test-integration`` was using the first positional arg passed to ``WHAT`` to filter the list of integration test packages.  This PR switches to passing ``WHAT`` verbatim to be consistent with how ``make test`` works.  That means the new way to scope execution to a single integration package will be:

```bash
make test-integration WHAT="./test/integration/auth" KUBE_TEST_ARGS="-run=^TestKindAuthorization$"
```

Instead of:

```bash
make test-integration WHAT="auth -test.run=^TestKindAuthorization$"
```

This PR also ensures that the script exits after running a single test case and that etcd cleanup is not done twice at the end of a successful test run.  Both were issues encountered while diagnosing the scoping issue.

cc: @thockin @deads2k @stevekuznetsov @ncdc @derekwaynecarr

### CODE REVIEW

File: hack/make-rules/test-integration.sh [hunk 1]
```diff
@@ -41,27 +41,33 @@ KUBE_TEST_ARGS=${KUBE_TEST_ARGS:-}
 kube::test::find_integration_test_dirs() {
   (
     cd ${KUBE_ROOT}
-    find test/integration/${1-} -name '*_test.go' -print0 \
+    find test/integration/ -name '*_test.go' -print0 \
```
Comment (stevekuznetsov) [inline hack/make-rules/test-integration.sh]: The `-print0` behavior is lost by the time this pipe hits the `sed` so it's functionally useless.

File: hack/make-rules/test-integration.sh [hunk 2]
```diff
@@ -41,27 +41,33 @@ KUBE_TEST_ARGS=${KUBE_TEST_ARGS:-}
 kube::test::find_integration_test_dirs() {
   (
```
Comment (stevekuznetsov) [inline hack/make-rules/test-integration.sh]: Seems weird to have this in a subshell

File: hack/make-rules/test-integration.sh [hunk 3]
```diff
@@ -41,27 +41,33 @@ KUBE_TEST_ARGS=${KUBE_TEST_ARGS:-}
...
     cd ${KUBE_ROOT}
```
Comment (stevekuznetsov) [inline hack/make-rules/test-integration.sh]: Why not prefix paths with this in the  `find` and do the conversion from `$KUBE_ROOT` to `$KUBE_GO_PACKAGE` in the post-process?

File: hack/make-rules/test-integration.sh [hunk 4]
```diff
@@ -41,27 +41,33 @@ KUBE_TEST_ARGS=${KUBE_TEST_ARGS:-}
...
   kube::log::status "Starting etcd instance"
+  CLEANUP_REQUIRED=1
   kube::etcd::start
   kube::log::status "Running integration test cases"
 
   # TODO: Re-enable race detection when we switch to a thread-safe etcd client
   # KUBE_RACE="-race"
   make -C "${KUBE_ROOT}" test \
-      WHAT="$(kube::test::find_integration_test_dirs ${2-} | paste -sd' ' -) $(echo ${@:3})" \
+      WHAT="${WHAT:-$(kube::test::find_integration_test_dirs | paste -sd' ' -)}" \
```
Comment (stevekuznetsov) [inline hack/make-rules/test-integration.sh]: What is the `paste` accomplishing here?

File: hack/make-rules/test-integration.sh [hunk 5]
```diff
@@ -87,22 +93,11 @@ trap cleanup EXIT
...
-    what_flags+=("${arg}")
-  fi
-done
-if [[ "${#what_flags[@]}" -eq 0 ]]; then
-  what_flags=''
-fi
-
-
 # Convert the CSV to an array of API versions to test
 IFS=';' read -a apiVersions <<< "${KUBE_TEST_API_VERSIONS}"
```
Comment (stevekuznetsov) [inline hack/make-rules/test-integration.sh]: You don't really need to convert to an array here, the `for` will honor `$IFS` as is:
```sh
IFS=';'
for apiVersion in ${KUBE_TEST_API_VERSIONS}; do
```

File: hack/make-rules/test-integration.sh (hunk 2 above)
Comment (marun) [inline hack/make-rules/test-integration.sh]: Uh, I didn't change this.  Not sure I want to tackle cleaning this up more than I already have.

File: hack/make-rules/test-integration.sh (hunk 3 above)
Comment (marun) [inline hack/make-rules/test-integration.sh]: ditto

**Comment by marun** [comment]
@stevekuznetsov I don't intend to take over maintenance of the script - and I didn't write it in the first place - so I'm not sure there's value in critiquing stuff I haven't changed.

File: hack/make-rules/test-integration.sh (hunk 2 above)
Comment (stevekuznetsov) [inline hack/make-rules/test-integration.sh]: Looks like my top-level comment got lost, sorry -- the changes look fine, some weirdness in the stuff around them from pre-existing code, not sure if you want to touch it or not.

Comment (marun) [inline hack/make-rules/test-integration.sh]: I agree that this script could use some cleanup, but for now I only have time to make it consistent.

**Comment by deads2k** [comment]
Can you provide the corresponding `hack/test-integration.sh` invocation?

**Comment by marun** [comment]
@deads2k 

```bash
KUBE_TEST_ARGS="-run=^TestKindAuthorization$" hack/test-integration.sh ./test/integration/auth
```

Unfortunately it would be necessary to use ``GOFLAGS`` instead of ``KUBE_GOFLAGS`` when calling ``hack/test-integration.sh`` due to the makefile setting ``KUBE_GOFLAGS`` to ``GOFLAGS``.

Is there a reason (other than muscle memory) to use ``hack/test-integration.sh`` given that it now wraps  ``make test-integration``?

**Comment by marun** [comment]
@k8s-bot cross build this

**Comment by deads2k** [comment]
> Is there a reason (other than muscle memory) to use hack/test-integration.sh given that it now wraps make test-integration?

It exists and this breaks reasonable parity.  If you're going to change it that significantly, just delete it or disable it completely.

**Comment by marun** [comment]
@deads2k ``hack/test-integration.sh`` is effectively deprecated already, as per the output when it's invoked.  I'm assuming there is a plan to remove it eventually but my goal here is to ensure that the documented way to run integration tests to run matches up with reality.

**Comment by deads2k** [comment]
> @deads2k hack/test-integration.sh is effectively deprecated already, as per the output when it's invoked. I'm assuming there is a plan to remove it eventually but my goal here is to ensure that the documented way to run integration tests to run matches up with reality.

Have it write a recommended command and just exit 1.  And have the recommended command be based on what the currently support syntax is.

File: hack/make-rules/test-integration.sh (hunk 4 above)
Comment (thockin) [inline hack/make-rules/test-integration.sh]: joining lines

**Comment by marun** [comment]
@deads2k Done

**Comment by marun** [comment]
@k8s-bot gci gke e2e test this

**Comment by marun** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by marun** [comment]
@deads2k bump

**Comment by marun** [comment]
@thockin I'd like to see this merge so the behavior can match the docs.

**Comment by marun** [comment]
@stevekuznetsov Can you please lgtm?

**Comment by spiffxp** [comment]
@ixdy maybe you can help if nobody else is being responsive here?  I'd like to see this get in too

**Review by stevekuznetsov — APPROVED** [review]
Didn't realize k8s CI cared about me. This looks fine.

**Comment by fejta** [comment]
/assign
/assign @stevekuznetsov @thockin @spxtr @janetkuo 

Various people who seem to care about integration and/or make files -- any more thoughts here? Can we can an lgtm?

**Comment by spxtr** [comment]
@k8s-bot bazel test this

**Comment by fejta** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by ixdy** [comment]
@k8s-bot gci gce e2e test this

**Pull request merged**